File: Kotlin/Encryption.kt
/*
 * Copyright (c) 2021, Shashank Verma <shashank.verma2002@gmail.com>(shank03)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 */

import kotlin.math.pow

@ExperimentalUnsignedTypes
object Encryption {

    private const val LINE_SEPARATOR = "--------------------------------------------------------------------- "

    private fun compress(input: String): HashMap<Char, ArrayList<ULong>> {
        val map = HashMap<Char, ArrayList<ULong>>()
        for (i in input.indices) {
            if (map.containsKey(input[i])) {
                map[input[i]]?.add((i + 1).toULong())
            } else {
                val pos = ArrayList<ULong>().also { it.add((i + 1).toULong()) }
                map[input[i]] = pos
            }
        }
        return map
    }

    private fun binaryToULong(bin: String): ULong {
        var out: ULong = 0U
        for (i in bin.indices) {
            if (bin[bin.length - 1 - i] == '1') out += 2.0.pow(i.toDouble()).toULong()
        }
        return out
    }

    private fun formatULong(i: ULong): String = String.format("%64s", i.toString(2)).replace(" ", "0")

    fun encrypt(text: String, key: String): String {
        var out = ""

        var keyMask: ULong = 0U
        for (c: Char in key) {
            keyMask += c.toInt().toULong()
        }

        val tLines = text.split("\n")
        for (line: String in tLines) {
            val data = compress(line)
            val info = line.length.toULong().xor(keyMask)
            out += LINE_SEPARATOR + "\n" + formatULong(info) + " \n"

            for ((_key, value) in data) {
                out += when (_key) {
                    '\n', '\r' -> "nl "
                    '\t' -> "tb "
                    ' ' -> "sp "
                    else -> "$_key "
                }

                var ps: ULong = 0U
                var c: ULong = 0U
                for (pos: ULong in value) {
                    ps = ps.or(pos.shl(8 * c.toInt()))
                    c++
                    if (c.toUInt() == 4U) {
                        ps = ps.or(c.shl(32))
                        out += formatULong(ps.xor(keyMask)) + " "
                        ps = 0U
                        c = 0U
                    }
                }
                if (ps.toUInt() != 0U) {
                    ps = ps.or(c.shl(32))
                }
                out += formatULong(ps.xor(keyMask)) + " \n"
            }
        }
        return out + LINE_SEPARATOR + "\n"
    }

    fun decrypt(input: String, key: String): String {
        var out = ""

        var keyMask: ULong = 0U
        for (c: Char in key) {
            keyMask += c.toInt().toULong()
        }

        val tLines = input.split("\n")
        var intoData = false
        var datLen = 0
        var data: CharArray? = null
        var char = '-'
        for (line: String in tLines) {
            if (line == LINE_SEPARATOR) {
                intoData = true
                if (data != null && datLen != 0) {
                    for (c: Char in data) out += c
                    out += "\n"
                }
                continue
            }
            if (intoData) {
                val len = line.trim().replace("\n", "")
                datLen = binaryToULong(len).xor(keyMask).toInt()
                data = CharArray(datLen)
                intoData = false
            } else {
                var count = 0
                val tTokens = line.replace("\n", "").split(" ")
                for (_token: String in tTokens) {
                    val token = _token.trim()
                    if (token == "") continue
                    if (count == 0) {
                        if (token.length == 1) char = token[0]
                        if (token.length == 2) {
                            if (token == "nl") char = '\n'
                            if (token == "sp") char = ' '
                            if (token == "tb") char = '\t'
                        }
                        count++
                    } else {
                        if (token.length == 64) {
                            val o = binaryToULong(token).xor(keyMask)
                            val numPos = o.shr(32)
                            var i = 0
                            while (i < numPos.toInt()) {
                                val pos = o.shr(8 * i).and(0xFFU)
                                if (data != null) data[pos.toInt() - 1] = char
                                i++
                            }
                        }
                    }
                }
            }
        }
        return out
    }
}

File: C++/encryption.cpp
/*
 * Copyright (c) 2021, Shashank Verma <shashank.verma2002@gmail.com>(shank03)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 */

#include "encryption.h"
#include <bitset>
#include <cstring>
#include <sstream>

void encryption::compress(const std::string &input, std::map<char, std::vector<uint64_t>> *map) {
    int len = input.length();
    for (int i = 0; i < len; i++) {
        if (map->find(input[i]) != map->end()) {    // char already exists
            (*map)[input[i]].push_back(i + 1);      // append the position in array
        } else {
            std::vector<uint64_t> pos;
            pos.push_back(i + 1);
            (*map)[input[i]] = pos;    // init map for char [i]
        }
    }
}

uint64_t encryption::binary_to_uint(const std::string &bin) {
    uint64_t out = 0;
    std::istringstream iss(bin);
    iss >> out;
    return out;
}

std::string encryption::encrypt(const std::string &text, const std::string &key) {
    std::stringstream outStr;

    // Key
    uint64_t keyMask = 0;
    for (char i : key) keyMask += keyMask ^ i;

    std::string t_token;
    size_t t_pos, t_index = 0;
    while ((t_pos = text.find('\n', t_index)) != std::string::npos) {
        t_token = text.substr(t_index, t_pos - t_index);

        // Mapping data
        std::map<char, std::vector<uint64_t>> data;
        encryption::compress(t_token, &data);

        uint64_t out = t_token.length() ^ keyMask;
        outStr << LINE_SEPARATOR << "\n"
               << (out) << " \n";

        // Encrypting mapped data
        for (auto &it : data) {
            outStr << ((uint64_t) it.first ^ keyMask) << " ";

            uint64_t ps = 0, c = 0;
            for (uint64_t i : it.second) {
                ps |= i << (8 * c), c++;
                if (c == 4) {
                    ps |= (uint64_t) 4 << 32;
                    outStr << (ps ^ keyMask) << " ";
                    ps = 0, c = 0;
                }
            }
            if (ps != 0) {
                ps |= c << 32;
            }
            outStr << (ps ^ keyMask) << " " << c << " \n";
        }
        t_index = t_pos + 1;
    }
    outStr << LINE_SEPARATOR << "\n";
    return outStr.str();
}

std::string encryption::decrypt(const std::string &encDat, const std::string &key) {
    std::stringstream outStr;

    // Key
    uint64_t keyMask = 0;
    for (char i : key) keyMask += keyMask ^ i;

    bool intoData = false;
    uint64_t datLength = 0;
    char *data = nullptr, c;
    std::string t_token;
    size_t t_pos, t_index = 0;
    while ((t_pos = encDat.find('\n', t_index)) != std::string::npos) {
        t_token = encDat.substr(t_index, t_pos - t_index);
        if (t_token == LINE_SEPARATOR) {
            intoData = true;
            if (data != nullptr && datLength != 0) {
                for (uint64_t i = 0; i < datLength; i++) outStr << data[i];
                outStr << "\n";
            }
            t_index = t_pos + 1;
            continue;
        }
        if (intoData) {
            std::string lenStr = t_token.substr(0, 64);
            datLength = binary_to_uint(lenStr) ^ keyMask;
            data = (char *) malloc(datLength * sizeof(char));
            intoData = false;
        } else {
            size_t p_pos, p_index = 0, count = 0;
            std::string p_token;
            while ((p_pos = t_token.find(' ', p_index)) != std::string::npos) {
                p_token = t_token.substr(p_index, p_pos - p_index);
                if (p_token.empty()) {
                    p_index = p_pos + 1;
                    continue;
                }
                uint64_t o = binary_to_uint(p_token) ^ keyMask;
                if (count == 0) {
                    c = o;
                    count++;
                } else {
                    uint64_t numPos = o >> 32;
                    for (uint64_t p = 0; p < numPos; p++) {
                        uint64_t pos = (o >> (8 * p)) & 0xFF;
                        if (data != nullptr) {
                            data[pos - 1] = c;
                        }
                    }
                }
                p_index = p_pos + 1;
            }
        }
        t_index = t_pos + 1;
    }
    return outStr.str();
}
